opentelemetry-auto-instrumentation/src/opentelemetry/auto_instrumentation/patcher.py

```diff
@@ -0,0 +1,76 @@
+# Copyright 2020, OpenTelemetry Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# type: ignore
+
+"""
+OpenTelemetry Auto Instrumentation Patcher
+"""
+
+from abc import ABC, abstractmethod
+from logging import getLogger
+
+_LOG = getLogger(__name__)
+
+
+class BasePatcher(ABC):
+    """An ABC for patchers"""
+
+    @abstractmethod
+    def _patch(self) -> None:
+        """Patch"""
+
+    @abstractmethod
+    def _unpatch(self) -> None:
+        """Unpatch"""
+
+    def patch(self) -> None:
+        """Patch"""
+
+        if not hasattr(self, "_is_patched"):
```

Comment:
Why not just put this in `__init__`?

Reply:
`BasePatcher` is an `ABC` and is expected to be subclassed. If one of these subclasses does not call `super` from its `__init__` then a weird bug arises. On the other hand, it is the usual thing to put this kind of attributes in `__init__` and should calling `super` be the responsibility of the subclasser? Very likely, yes. I don't have very strong opinions on this one, so I can change it.

Reply:
After some more consideration, I have just added it to `__init__`. I think it is expected that `super` be called.